File: Algos & templates/题目/动态规划/数位/LC1012.py
class Solution:
    def numDupDigitsAtMostN(self, n: int) -> int:
        def cal(n: int):
            s = str(n)
            @cache
            def dfs(i: int, sta: int, islimit: bool, isnum: bool) -> int:
                # isnum: 前边是否填了数 islimit: 是否受到n 的限制
                if (i == len(s)):
                    return int(isnum)
                res = 0
                if (not isnum):
                    res += dfs(i + 1, sta, False, False)
                low = 0 if (isnum) else 1
                up = int(s[i]) if islimit else 9
                for d in range(low, up + 1):
                    if (((sta >> d) & 1) == 0):
                        f = dfs(i + 1, sta | (1 << d), islimit and d == up, True)
                        res += f
                return res

            return dfs(0, 0, True, False)

        return n - cal(n)

File: Algos & templates/题目/动态规划/数位/LC1012.cpp
class Solution {
public:
    int numDupDigitsAtMostN(int n) {
        string s = to_string(n);
        int m = s.size();
        int dp[m + 1][1 << 10];
        memset(dp, -1, sizeof(dp));
        //是否收到 n 的限制, 前一位是否填了数
        function<int(int, int, bool, bool)>dfs = [&](int i, int sta, bool islimit, bool isnum) {
            if (i == m) {
                return (int)isnum;
            }
            if (dp[i][sta] != -1 and not islimit and isnum)return dp[i][sta];
            int res = 0;
            if (not isnum)res += dfs(i + 1, sta, false, false);
            int low = isnum ? 0 : 1;
            int up = islimit ? s[i] - '0' : 9;
            for (int d = low; d <= up; d++) {
                if (((sta >> d) & 1) == 0) {
                    int ret = dfs(i + 1, sta | (1 << d), islimit and d == up, true);
                    res += ret;
                }
            }
            if (not islimit and isnum)
                dp[i][sta] = res;
            return res;
        };
        return n - dfs(0, 0, true, false);
    }
};